File: Tests/vs_cpp/vs_cpp/CS.cpp
//#include "precompile.h"
# include "CS.h"
# include "xxlib.h"
using namespace xxlib;

const short ::CS::Ns2::Pkg2::packageId = 1;
const short ::CS::Ns1::Pkg1::packageId = 2;
const short ::CS::Ns1::Pkg2::packageId = 3;
const short ::CS::RootPkg1::packageId = 0;
void ::CS::Ns2::Pkg2::WriteTo( ByteBuffer& bb ) const
{
    bb.Write( this->re2 ); 
}

bool ::CS::Ns2::Pkg2::ReadFrom( ByteBuffer& bb )
{
    if( !bb.Read( this->re2 ) ) return false;
    return true;
}
void ::CS::Ns1::Pkg1::WriteTo( ByteBuffer& bb ) const
{
    bb.Write( this->ns1_enum1 ); 
}

bool ::CS::Ns1::Pkg1::ReadFrom( ByteBuffer& bb )
{
    if( !bb.Read( this->ns1_enum1 ) ) return false;
    return true;
}
void ::CS::Ns1::Pkg2::WriteTo( ByteBuffer& bb ) const
{
    bb.Write( this->re1 );
    bb.Write( this->ns1_pkg1 );
    bb.Write( this->root_ns2_pkg1 ); 
}

bool ::CS::Ns1::Pkg2::ReadFrom( ByteBuffer& bb )
{
    if( !bb.Read( this->re1 ) ) return false;
    if( !bb.Read( this->ns1_pkg1 ) ) return false;
    if( !bb.Read( this->root_ns2_pkg1 ) ) return false;
    return true;
}
void ::CS::RootPkg1::WriteTo( ByteBuffer& bb ) const
{
    bb.Write( this->m_root_e1 );
    bb.Write( this->m_ns1_pkg2 );
    bb.Write( this->m_byte );
    bb.Write( this->m_ushort );
    bb.VarWrite( this->m_uint );
    bb.VarWrite( this->m_ulong );
    bb.Write( this->m_sbyte );
    bb.Write( this->m_short );
    bb.VarWrite( this->m_int );
    bb.VarWrite( this->m_long );
    bb.Write( this->m_double );
    bb.Write( this->m_float );
    bb.Write( this->m_bool );
    bb.Write( this->m_DateTime );
    bb.Write( this->m_string );
    bb.Write( this->m_ByteBuffer );
    bb.Write( this->m_bytes );
    bb.Write( this->m_ns2enum1s );
    bb.Write( this->m_bools ); 
}

bool ::CS::RootPkg1::ReadFrom( ByteBuffer& bb )
{
    if( !bb.Read( this->m_root_e1 ) ) return false;
    if( !bb.Read( this->m_ns1_pkg2 ) ) return false;
    if( !bb.Read( this->m_byte ) ) return false;
    if( !bb.Read( this->m_ushort ) ) return false;
    if( !bb.VarRead( this->m_uint ) ) return false;
    if( !bb.VarRead( this->m_ulong ) ) return false;
    if( !bb.Read( this->m_sbyte ) ) return false;
    if( !bb.Read( this->m_short ) ) return false;
    if( !bb.VarRead( this->m_int ) ) return false;
    if( !bb.VarRead( this->m_long ) ) return false;
    if( !bb.Read( this->m_double ) ) return false;
    if( !bb.Read( this->m_float ) ) return false;
    if( !bb.Read( this->m_bool ) ) return false;
    if( !bb.Read( this->m_DateTime ) ) return false;
    if( !bb.Read( this->m_string, 4, 16 ) ) return false;
    if( !bb.Read( this->m_ByteBuffer, 0, 128 ) ) return false;
    if( !bb.Read( this->m_bytes ) ) return false;
    if( !bb.Read( this->m_ns2enum1s, 0, 0 ) ) return false;
    if( !bb.Read( this->m_bools, 0, 0 ) ) return false;
    return true;
}

File: Tests/vs_cs/vs_cs/CS.cs
using System;
using System.Collections.Generic;
using xxlib;
namespace CS
{
/// <summary>
/// RootEnum1's desc
/// </summary>
public enum Enum1 : byte
{
    /// <summary>
    /// RootEnum1 Item1's desc
    /// </summary>
    Item1 = 0, 
    /// <summary>
    /// RootEnum1 Ite2's desc
    /// </summary>
    Item2 = 1, 
    Item3 = 100, 
    Item4 = 101, 
    Item5 = 102, 
};
namespace Ns2
{
public enum Enum1 : byte
{
    item1 = 0, 
};
} // Ns2
namespace Ns1
{
public enum Enum1 : ushort
{
    item1 = 0, 
};
} // Ns1
partial class RootPkg1
{
    public const short packageId = 0;

    public global::CS.Enum1 m_root_e1 = global::CS.Enum1.Item1;
    public global::CS.Ns1.Pkg2 m_ns1_pkg2 = new global::CS.Ns1.Pkg2();
    public byte m_byte = 0;
    public ushort m_ushort = 0;
    public uint m_uint = 0;
    public ulong m_ulong = 0;
    public sbyte m_sbyte = 0;
    public short m_short = 0;
    public int m_int = 0;
    public long m_long = 0;
    public double m_double = 0;
    public float m_float = 0;
    public bool m_bool = false;
    public DateTime m_DateTime = new DateTime( 1991, 1, 1, 0, 0, 0 );
    public string m_string = "";
    public ByteBuffer m_ByteBuffer = new ByteBuffer();
    public byte[] m_bytes = new byte[ 8 ];
    public List<global::CS.Ns2.Enum1> m_ns2enum1s = new List<global::CS.Ns2.Enum1>();
    public List<bool> m_bools = new List<bool>();
};
namespace Ns2
{
partial class Pkg2
{
    public const short packageId = 1;

    public global::CS.Enum1 re2 = global::CS.Enum1.Item1;
};
} // Ns2
namespace Ns1
{
partial class Pkg1
{
    public const short packageId = 2;

    public global::CS.Ns1.Enum1 ns1_enum1 = global::CS.Ns1.Enum1.item1;
};
} // Ns1
namespace Ns1
{
partial class Pkg2
{
    public const short packageId = 3;

    public global::CS.Enum1 re1 = global::CS.Enum1.Item1;
    public global::CS.Ns1.Pkg1 ns1_pkg1 = new global::CS.Ns1.Pkg1();
    public global::CS.Ns2.Pkg2 root_ns2_pkg1 = new global::CS.Ns2.Pkg2();
};
} // Ns1
} // CS
